### DISCUSSION TITLE
Android widgets

### DISCUSSION THREAD

**Discussion by moolight-seashell** [slint-ui/slint#6186] [Q&A]
Well, this question may be absurd, as android widgets are specific to android and java.
In any case, widgets seem rather limited in what you can do with them.

But could we even imagine creating android widgets via slint in rust?

eve if i d'ont know how we can do an generic jni things for AppWidgetProvider or even for RemoteViews ?
maybe a generic Image widgets on when slint is rendering ? ( probably extrely slow)

but on the other hand, is it even possible to integrate an android / java widget into a rust project?

**Comment by ogoffart** [comment]
I'm not so familiar with the details of android but I hope it will be possible, although currently some integration code will be needed.

Right now the android backend uses the NativeActivity behind the scene. (It could also use GameActivity if needed)
So it all depends if we can put these activity on a widget.

Some glue code in java will be needed, but it is possible to call rust from java and java from rust (using the jni crate)

**Reply by smir-ant** [reply]
what about now? want to easy and comfort work with home screen widgets. android first, but and ios perspective. i have great idea and mvp app, but jni and kotlin... painfull

**Reply by moolight-seashell** [reply]
from what **I've understand** Android AppWidgets use RemoteViews,
which are essentially predefined android views (using the native android interface system).
basically the system reads and renders xml native android interface directly

so there's almost no chance that this will ever be possible.
The most “realistic” would be an android widget loading slint rendered images.

not clean and probably not very relevant.

I'd advise you to start from the principle that android widgets will only be possible if you do the widget part in a separate pure android app (and comunicate the data with intents using jni a c abi, for example).

**Reply by ogoffart** [reply]
We probably could get it work if we use the android-view crate  in the future.
https://github.com/rust-mobile/android-view
